File: packages/e2e-test/src/__tests__/env.test.ts
/**
 * End-to-End Tests for the env command.
 *
 * These tests verify that the env command correctly:
 * - Creates virtual environment directory with config.ts, source, source.ps1, tools/
 * - Deletes virtual environment with --force flag
 * - Produces correct activation scripts
 */
import { afterAll, beforeAll, describe, expect, it } from 'bun:test';
// oxlint-disable-next-line import/no-unassigned-import
import '@dotfiles/testing-helpers';
import { dedentString } from '@dotfiles/cli';
import { Architecture, Platform } from '@dotfiles/core';
import fs from 'node:fs';
import path from 'node:path';
import { TestHarness } from './helpers/TestHarness';

describe('E2E: env command', () => {
  const platformConfigs: ReadonlyArray<{
    platform: Platform;
    architecture: Architecture;
    name: string;
  }> = [
    { platform: Platform.MacOS, architecture: Architecture.Arm64, name: 'macOS ARM64' },
    { platform: Platform.Linux, architecture: Architecture.X86_64, name: 'Linux x86_64' },
  ];

  for (const config of platformConfigs) {
    describe(`${config.name}`, () => {
      const harness: TestHarness = new TestHarness({
        testDir: import.meta.dir,
        configPath: 'fixtures/main/config.ts',
        platform: config.platform,
        architecture: config.architecture,
      });

      const testEnvName = 'test-env';
      const testEnvDir = path.join(import.meta.dir, testEnvName);
      const defaultEnvDir = path.join(import.meta.dir, 'env');
      // Use a temp directory for dynamically created envs, separate from static fixtures
      const activatedEnvTempDir = path.join(import.meta.dir, 'activated-env-temp');

      afterAll(async () => {
        await Promise.all([
          fs.promises.rm(testEnvDir, { recursive: true, force: true }),
          fs.promises.rm(defaultEnvDir, { recursive: true, force: true }),
          fs.promises.rm(activatedEnvTempDir, { recursive: true, force: true }),
        ]);
      });

      describe('env create', () => {
        beforeAll(async () => {
          await fs.promises.rm(testEnvDir, { recursive: true, force: true });
          await harness.runCommand(['env', 'create', testEnvName], {
            env: { DOTFILES_BUILT_PACKAGE_NAME: '@dotfiles/cli' },
          });
        });

        it('should create config.ts file with defineConfig import', async () => {
          const configPath = path.join(testEnvDir, 'config.ts');
          const exists = await fs.promises.access(configPath).then(() => true).catch(() => false);
          expect(exists).toBe(true);

          const content = await fs.promises.readFile(configPath, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            import { defineConfig } from ${expect.anything};
          `;
        });

        it('should create POSIX source script with shebang', async () => {
          const sourcePath = path.join(testEnvDir, 'source');
          const exists = await fs.promises.access(sourcePath).then(() => true).catch(() => false);
          expect(exists).toBe(true);

          const content = await fs.promises.readFile(sourcePath, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            #!/bin/sh
            # ====${expect.anything}
          `;
        });

        it('should create PowerShell source script with header', async () => {
          const sourcePath = path.join(testEnvDir, 'source.ps1');
          const exists = await fs.promises.access(sourcePath).then(() => true).catch(() => false);
          expect(exists).toBe(true);

          const content = await fs.promises.readFile(sourcePath, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            # ====${expect.anything}
            # Dotfiles Virtual Environment Activation Script (PowerShell)
          `;
        });

        it('should create tools directory', async () => {
          const toolsDir = path.join(testEnvDir, 'tools');
          const stat = await fs.promises.stat(toolsDir);
          expect(stat.isDirectory()).toBe(true);
        });

        it('should make POSIX source script executable', async () => {
          const sourcePath = path.join(testEnvDir, 'source');
          const stat = await fs.promises.stat(sourcePath);
          const isExecutable = (stat.mode & 0o111) !== 0;
          expect(isExecutable).toBe(true);
        });

        it('should export XDG_CONFIG_HOME in source script', async () => {
          const sourcePath = path.join(testEnvDir, 'source');
          const content = await fs.promises.readFile(sourcePath, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            export XDG_CONFIG_HOME="\${_dotfiles_script_dir}/.config"
          `;
        });

        it('should export XDG_CONFIG_HOME in PowerShell script', async () => {
          const sourcePath = path.join(testEnvDir, 'source.ps1');
          const content = await fs.promises.readFile(sourcePath, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            $env:XDG_CONFIG_HOME = "$scriptDir\\.config"
          `;
        });
      });

      describe('env create when already exists', () => {
        beforeAll(async () => {
          await fs.promises.rm(testEnvDir, { recursive: true, force: true });
          await harness.runCommand(['env', 'create', testEnvName], {
            env: { DOTFILES_BUILT_PACKAGE_NAME: '@dotfiles/cli' },
          });
        });

        it('should fail when environment already exists', async () => {
          const result = await harness.runCommand(['env', 'create', testEnvName]);

          expect(result.code).not.toBe(0);
          expect(result.stdout).toMatchLooseInlineSnapshot`
            WARN	Environment already exists at ${expect.anything}
            ERROR	Environment create failed: Environment already exists at ${expect.anything}
          `;
        });
      });

      describe('env delete', () => {
        beforeAll(async () => {
          await fs.promises.rm(testEnvDir, { recursive: true, force: true });
          await harness.runCommand(['env', 'create', testEnvName], {
            env: { DOTFILES_BUILT_PACKAGE_NAME: '@dotfiles/cli' },
          });
          await harness.runCommand(['env', 'delete', testEnvName, '--force']);
        });

        it('should have removed the environment directory', async () => {
          const exists = await fs.promises.access(testEnvDir).then(() => true).catch(() => false);
          expect(exists).toBe(false);
        });
      });

      describe('env delete non-existent', () => {
        it('should fail when environment does not exist', async () => {
          const result = await harness.runCommand(['env', 'delete', 'non-existent-env', '--force']);

          expect(result.code).not.toBe(0);
          expect(result.stdout).toMatchLooseInlineSnapshot`
            ERROR	Environment not found at ${expect.anything}
          `;
        });
      });

      describe('env create with default name', () => {
        beforeAll(async () => {
          await fs.promises.rm(defaultEnvDir, { recursive: true, force: true });
          await harness.runCommand(['env', 'create'], {
            env: { DOTFILES_BUILT_PACKAGE_NAME: '@dotfiles/cli' },
          });
        });

        it('should have created the default env directory', async () => {
          const exists = await fs.promises.access(defaultEnvDir).then(() => true).catch(() => false);
          expect(exists).toBe(true);
        });
      });

      describe('env delete with default name', () => {
        beforeAll(async () => {
          await fs.promises.rm(defaultEnvDir, { recursive: true, force: true });
          await harness.runCommand(['env', 'create'], {
            env: { DOTFILES_BUILT_PACKAGE_NAME: '@dotfiles/cli' },
          });
          await harness.runCommand(['env', 'delete', '--force']);
        });

        it('should have removed the default env directory', async () => {
          const exists = await fs.promises.access(defaultEnvDir).then(() => true).catch(() => false);
          expect(exists).toBe(false);
        });
      });

      describe('activated environment with tool', () => {
        const shellOnlyToolContent = dedentString(`
          import { defineTool } from '@dotfiles/cli';

          export default defineTool((install) =>
            install()
              .zsh((shell) =>
                shell.aliases({ 'env-test-alias': 'echo "env tool works"' })
              )
              .bash((shell) =>
                shell.aliases({ 'env-test-alias': 'echo "env tool works"' })
              )
          );
        `);

        beforeAll(async () => {
          await fs.promises.rm(activatedEnvTempDir, { recursive: true, force: true });
          await harness.runCommand(['env', 'create', 'activated-env-temp'], {
            env: { DOTFILES_BUILT_PACKAGE_NAME: '@dotfiles/cli' },
          });
          // Add a shell-only tool that doesn't require network
          const toolPath = path.join(activatedEnvTempDir, 'tools', 'test-tool.tool.ts');
          await fs.promises.writeFile(toolPath, shellOnlyToolContent);
        });

        it('should use env config when env is activated', async () => {
          const result = await harness.runCommandWithActivatedEnv(activatedEnvTempDir, ['generate']);

          expect(result.code).toBe(0);
        });

        it('should create .generated directory in env when activated', async () => {
          const generatedDir = path.join(activatedEnvTempDir, '.generated');
          const exists = await fs.promises.access(generatedDir).then(() => true).catch(() => false);
          expect(exists).toBe(true);
        });

        it('should create shell-scripts in env .generated directory', async () => {
          const shellScriptsDir = path.join(activatedEnvTempDir, '.generated', 'shell-scripts');
          const exists = await fs.promises.access(shellScriptsDir).then(() => true).catch(() => false);
          expect(exists).toBe(true);
        });

        it('should include tool alias in generated zsh script', async () => {
          const zshScript = path.join(activatedEnvTempDir, '.generated', 'shell-scripts', 'main.zsh');
          const content = await fs.promises.readFile(zshScript, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            alias env-test-alias="echo \\"env tool works\\""
          `;
        });

        it('should include tool alias in generated bash script', async () => {
          const bashScript = path.join(activatedEnvTempDir, '.generated', 'shell-scripts', 'main.bash');
          const content = await fs.promises.readFile(bashScript, 'utf8');
          expect(content).toMatchLooseInlineSnapshot`
            alias env-test-alias="echo \\"env tool works\\""
          `;
        });
      });
    });
  }
});

File: tests/e2e/env_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/alexgorbatchev/dotfiles/pkg/fs"
	"github.com/alexgorbatchev/dotfiles/pkg/venv"
)

func TestE2EVenv(t *testing.T) {
	t.Parallel()

	t.Run("venv manager operations", func(t *testing.T) {
		tempDir := t.TempDir()
		osFS := fs.NewOSFS()
		mgr := venv.NewManager(osFS)

		testEnvName := "test-env"
		testEnvDir := filepath.Join(tempDir, testEnvName)

		// Create
		info, err := mgr.Create(tempDir, testEnvName, false)
		if err != nil {
			t.Fatalf("failed to create env: %v", err)
		}
		if info.EnvDir != testEnvDir {
			t.Errorf("expected env dir %q, got %q", testEnvDir, info.EnvDir)
		}

		// Verify files
		configPath := filepath.Join(testEnvDir, "dotfiles.config.ts")
		sourcePath := filepath.Join(testEnvDir, "source")
		psPath := filepath.Join(testEnvDir, "source.ps1")
		toolsDir := filepath.Join(testEnvDir, "tools")

		if _, err := os.Stat(configPath); err != nil {
			t.Errorf("config file does not exist: %v", err)
		}
		if _, err := os.Stat(sourcePath); err != nil {
			t.Errorf("POSIX source script does not exist: %v", err)
		}
		if _, err := os.Stat(psPath); err != nil {
			t.Errorf("PowerShell source script does not exist: %v", err)
		}
		fi, err := os.Stat(toolsDir)
		if err != nil || !fi.IsDir() {
			t.Errorf("tools directory does not exist or is not a directory")
		}

		// Verify script permissions
		sfi, err := os.Stat(sourcePath)
		if err != nil {
			t.Fatalf("failed to stat source path: %v", err)
		}
		if sfi.Mode()&0111 == 0 {
			t.Errorf("expected POSIX source script to be executable")
		}

		// Check POSIX source content
		srcContentBytes, err := os.ReadFile(sourcePath)
		if err != nil {
			t.Fatalf("failed to read source path: %v", err)
		}
		srcContent := string(srcContentBytes)
		if !strings.HasPrefix(srcContent, "#!/bin/sh") {
			t.Errorf("expected POSIX source to start with shebang")
		}
		if !strings.Contains(srcContent, "export XDG_CONFIG_HOME=\"${_dotfiles_script_dir}/.config\"") {
			t.Errorf("expected POSIX source to export XDG_CONFIG_HOME")
		}

		// Check PowerShell source content
		psContentBytes, err := os.ReadFile(psPath)
		if err != nil {
			t.Fatalf("failed to read ps path: %v", err)
		}
		psContent := string(psContentBytes)
		if !strings.Contains(psContent, "Dotfiles Virtual Environment Activation Script (PowerShell)") {
			t.Errorf("expected PS source to have PS comment header")
		}
		if !strings.Contains(psContent, "$env:XDG_CONFIG_HOME = \"$targetDir\\.config\"") {
			t.Errorf("expected PS source to export XDG_CONFIG_HOME")
		}

		// IsValidEnv check
		isValid, err := mgr.IsValidEnv(testEnvDir)
		if err != nil || !isValid {
			t.Errorf("IsValidEnv returned %v, err: %v, expected true", isValid, err)
		}

		// Create when already exists error
		_, err = mgr.Create(tempDir, testEnvName, false)
		if err == nil {
			t.Error("expected error creating already existing env without force")
		}

		// Create when already exists with force success
		_, err = mgr.Create(tempDir, testEnvName, true)
		if err != nil {
			t.Errorf("expected success creating already existing env with force: %v", err)
		}

		// Delete
		err = mgr.Delete(testEnvDir)
		if err != nil {
			t.Fatalf("failed to delete env: %v", err)
		}

		if _, err := os.Stat(testEnvDir); !os.IsNotExist(err) {
			t.Error("expected env directory to be removed")
		}

		// Delete non-existent
		err = mgr.Delete(filepath.Join(tempDir, "non-existent-env"))
		if err == nil {
			t.Error("expected error when deleting non-existent env")
		}
	})

	t.Run("activated environment with tool generation", func(t *testing.T) {
		h := NewTestHarness(t, HarnessOptions{})

		// Create virtual environment under temp dir
		osFS := fs.NewOSFS()
		mgr := venv.NewManager(osFS)
		envName := "activated-env-temp"
		info, err := mgr.Create(h.TempDir, envName, true)
		if err != nil {
			t.Fatalf("failed to create virtual environment: %v", err)
		}

		// Customize config.ts inside the virtual-env to define paths
		// Import from @alexgorbatchev/dotfiles directly so esbuild resolves it properly
		venvConfigContent := `export default {
  paths: {
    generatedDir: "` + filepath.ToSlash(info.EnvDir) + `/.generated",
    homeDir: "{paths.generatedDir}/user-home",
    targetDir: "{paths.generatedDir}/user-bin",
    toolConfigsDir: "` + filepath.ToSlash(info.EnvDir) + `/tools",
  },
};
`
		err = os.WriteFile(filepath.Join(info.EnvDir, "config.ts"), []byte(venvConfigContent), 0644)
		if err != nil {
			t.Fatalf("failed to write config.ts: %v", err)
		}

		// Add a shell-only tool that doesn't require network
		shellOnlyToolContent := `import { defineTool } from "@alexgorbatchev/dotfiles";

export default defineTool((install) =>
  install()
    .zsh((shell) =>
      shell.aliases({ 'env-test-alias': 'echo "env tool works"' })
    )
    .bash((shell) =>
      shell.aliases({ 'env-test-alias': 'echo "env tool works"' })
    )
);
`
		err = os.WriteFile(filepath.Join(info.EnvDir, "tools", "test-tool.tool.ts"), []byte(shellOnlyToolContent), 0644)
		if err != nil {
			t.Fatalf("failed to write tool config: %v", err)
		}

		// Run generate with harness targeting this env's config
		h.ConfigPath = filepath.Join(info.EnvDir, "config.ts")
		stdout, stderr, exitCode, err := h.Generate()
		if err != nil {
			t.Fatalf("generate failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
		}
		if exitCode != 0 {
			t.Fatalf("generate returned exit code %d, stdout: %s\nstderr: %s", exitCode, stdout, stderr)
		}

		// Assert .generated directory exists
		generatedDir := filepath.Join(info.EnvDir, ".generated")
		if _, err := os.Stat(generatedDir); err != nil {
			t.Errorf("expected .generated directory to exist in env")
		}

		// Assert shell-scripts exists
		shellScriptsDir := filepath.Join(generatedDir, "shell-scripts")
		if _, err := os.Stat(shellScriptsDir); err != nil {
			t.Errorf("expected shell-scripts directory to exist in env .generated")
		}

		// Assert aliases in generated scripts
		zshScript := filepath.Join(shellScriptsDir, "main.zsh")
		zshContent, err := os.ReadFile(zshScript)
		if err != nil {
			t.Fatalf("failed to read main.zsh: %v", err)
		}
		if !strings.Contains(string(zshContent), "alias env-test-alias='echo \"env tool works\"'") &&
			!strings.Contains(string(zshContent), `alias env-test-alias="echo 'env tool works'"`) &&
			!strings.Contains(string(zshContent), `alias env-test-alias="echo \"env tool works\""`) {
			t.Errorf("expected main.zsh to contain env-test-alias, but got:\n%s", string(zshContent))
		}

		bashScript := filepath.Join(shellScriptsDir, "main.bash")
		bashContent, err := os.ReadFile(bashScript)
		if err != nil {
			t.Fatalf("failed to read main.bash: %v", err)
		}
		if !strings.Contains(string(bashContent), "alias env-test-alias='echo \"env tool works\"'") &&
			!strings.Contains(string(bashContent), `alias env-test-alias="echo 'env tool works'"`) &&
			!strings.Contains(string(bashContent), `alias env-test-alias="echo \"env tool works\""`) {
			t.Errorf("expected main.bash to contain env-test-alias, but got:\n%s", string(bashContent))
		}
	})
}
